From 7525402a80854f3bc74ed975e6f90ffbfadfd2ca Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:14:59 -0400
Subject: [net] udp: call __skb_checksum_complete when doing full checksum

Message-id: <1409652920-431-37-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90172
O-Subject: [PATCH 36/57] [RHEL7.1 net] udp: call __skb_checksum_complete when doing full checksum
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit bbdff225ede6527f91184b2a7903df8aad803ace
Author: Tom Herbert <therbert@google.com>
Date:   Sat Jun 14 23:24:20 2014 -0700

    udp: call __skb_checksum_complete when doing full checksum

    In __udp_lib_checksum_complete check if checksum is being done over all
    the data (len is equal to skb->len) and if it is call
    __skb_checksum_complete instead of __skb_checksum_complete_head. This
    allows checksum to be saved in checksum complete.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/net/udp.h b/include/net/udp.h
index 1a94f86..c8b5d65 100644
--- a/include/net/udp.h
+++ b/include/net/udp.h
@@ -111,7 +111,9 @@ struct sk_buff;
  */
 static inline __sum16 __udp_lib_checksum_complete(struct sk_buff *skb)
 {
- return __skb_checksum_complete_head(skb, UDP_SKB_CB(skb)->cscov);
+ return (UDP_SKB_CB(skb)->cscov == skb->len ?
+  __skb_checksum_complete(skb) :
+  __skb_checksum_complete_head(skb, UDP_SKB_CB(skb)->cscov));
 }
 
 static inline int udp_lib_checksum_complete(struct sk_buff *skb)
-- 
1.7.1